### ISSUE TITLE
Can't check if checksum verification has been successful

### ISSUE THREAD

**Issue by sergio-gimenez** [p4lang/p4c#2032]
I'm trying to add checksum verification for custom packets, but I can't really achieve it.

 This is my code in `MyVerifyChecksum` control block:

```p4
control MyVerifyChecksum(inout headers hdr,
                        inout metadata meta) {
    apply {
                verify_checksum(hdr.efcp.isValid(),
            { hdr.efcp.vers,
	      hdr.efcp.dstAddr,
              hdr.efcp.srcAddr,
              hdr.efcp.qosID,
              hdr.efcp.dstCEPID,
              hdr.efcp.srcCEPID,
              hdr.efcp.pduType,
              hdr.efcp.flags,
              hdr.efcp.length,
              hdr.efcp.seqnum },
            hdr.efcp.hdrChecksum,
            HashAlgorithm.csum16);
    }
}
```
So, following the instructions in [v1model.p4](https://github.com/p4lang/p4c/blob/master/p4include/v1model.p4) for the `verify_checksum` method:

> Verifies the checksum of the supplied data.  If this method detects
that a checksum of the data is not correct, then the value of the
standard_metadata checksum_error field will be equal to 1 when the
packet begins ingress processing.

So in the ingress processing control I check the state of the bit:

```p4
apply {
        if(hdr.efcp.isValid()){
            if(standard_metadata.checksum_error == 0)
                efcp_lpm.apply();
        }
}
```

But the point is that the `checksum_error` bit **never changes its state**.

Taking a look at the logs (`/tmp/p4.sx.log `) I have noticed that the checksum is correctly verified (somewhere).

* If the checksum is correct:

```
[08:41:50.983] [bmv2] [D] [thread 4711] [41.0] [cxt 0] Updating checksum 'cksum'
[08:41:50.983] [bmv2] [D] [thread 4711] [41.0] [cxt 0] Updating checksum 'cksum_0'
[08:41:54.649] [bmv2] [D] [thread 4709] [42.0] [cxt 0] Verifying checksum 'cksum': true
[08:41:54.649] [bmv2] [D] [thread 4709] [42.0] [cxt 0] Verifying checksum 'cksum_0': true
```

* If the checksum is incorrect:

```
[08:41:37.846] [bmv2] [D] [thread 4722] [32.0] [cxt 0] Verifying checksum 'cksum': false
[08:41:37.846] [bmv2] [E] [thread 4722] [32.0] [cxt 0] Checksum 'cksum' is not correct
[08:41:37.846] [bmv2] [D] [thread 4722] [32.0] [cxt 0] Verifying checksum 'cksum_0': false
[08:41:37.846] [bmv2] [E] [thread 4722] [32.0] [cxt 0] Checksum 'cksum_0' is not correct
```

In conclusion, the checksum is verified, but I don't know how to check if the verification
was successful. The documentation says that it should be in the `checksum_error` bit,
but it is not.

Any ideas about how check the verification properly?

**Comment by jafingerhut** [comment]
Do you know which version of the p4c compiler and behavioral-model simple_switch code you are using?

Do you have a full P4_16 program that can be compiled, and a packet that exhibits this problem?

There is a test program for p4c that exercises this behavior, and this test should be executed before any proposed change is made to p4c, and it exercises the path where `checksum_error` is assigned the value 1 (it is p4c/testdata/p4_16_samples/checksum2-bmv2.p4 and the corresponding .stf file), so unless those tests are not being run somehow, the feature should not be completely broken, but perhaps you have found a bug.

**Comment by sergio-gimenez** [comment]
Indeed, I was using an old version. I was using this [VM of 2018](https://p4.org/events/2018-06-06-p4-developer-day/), now I tried this [new VM](https://p4.org/events/2019-04-30-p4-developer-day/) and it seems to work.

Taking a look to the log of the switch, the `checksum_error` bit changes the value properly when a wrong checksum is detected:

```
[09:07:56.111] [bmv2] [T] [thread 8973] [28.0] [cxt 0] efcpRouter.p4(178) Condition "standard_metadata.checksum_error == 0" (node_3) is false
```
Thank you very much for your fast answer!

**Comment by jnfoster** [comment]
Looks like this can be closed?
